Support for IPv4 fragments

qmonnet opened the pull request:
This PR is work in progress and is submitted for reviewers to validate the approach. It contains the BPF-related changes. They were successfully compiled, but have not been otherwise tested yet. I am still working on the relevant parts in cilium-agent (creation of the BPF map) and on the tests. The two patches have a number of observations especially regarding the location of the changes in the code, please feel free to have a look and comment.

High-level description: To add support for IPv4 datagram fragments when doing L4 load balancer or L4/L3+L4 policy lookups, we need a way to retrieve the L4 ports for the datagram the fragment belongs to. We can avoid reassembling packets (and disassembling it again on egress) by adding another eBPF table to do the association between the datagram (“identification” field) and the L4 ports related to the datagram, i.e. that can be extracted from the first fragment. The table is used as follows:

- When the first fragment reaches the pipeline, fragmentation is detected (“identification”, “more fragments” flag). The 5-tuple is extracted as usual, but an update is also performed in the fragmentation table (key: src IP, dst IP, L4 proto, datagram id; value: L4 ports).

- When the following fragments reach the pipeline, fragmentation is detected too (“identification”, “fragment offset”). As L4 ports are not present, the regular key extraction is not performed. Instead, a lookup in the fragmentation table is realised for retrieving the L4 ports associated with the datagram. Once this information is known, the normal processing can resume (starting with the lookup in the conntrack table). Packets cross the pipeline without being reassembled/disassembled again.

IPv4 fragment supports would be enabled by default, with maybe a flag to disable it (`#ifdef IPV4_FRAGMENTS`). It would not be available on older kernels that do not support LRU maps (e.g. 4.9).

Fixes: #10076
(in part)

**Edit 2020-02-21** Now with the code to add the map in cilium-agent. Compiles, but not tested yet.
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10264)
<!-- Reviewable:end -->

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -22,6 +22,29 @@
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CILIUM_IPV4_FRAG_MAP_MAX_ENTRIES,
+	.flags		= CONDITIONAL_PREALLOC,
+};
```
If #10241 is merged first, you'll need to open this map from the cilium-agent, in `daemon.InitMaps()`. (If not, I can do it when I rebase.) To do that, you can have a look at ab4b45623cce7a77 ("maps, daemon: Create all global maps from daemon") and add similar code as I did for e.g. the `signalmap`. Note that you'll need to update the Makefile to generate the deepcopy file. There's already code in `daemon.InitMaps()` to detect if LRU hashmaps are supported.

qmonnet replied:
I'll definitely look at that and try to open it from cilium-agent directly, let's try to avoid a rebase :). Thanks for the heads up and pointers!

pchaigno commented on pkg/maps/fragmap/fragmap.go:
```diff
@@ -0,0 +1,107 @@
+type FragmentValue struct {
+	destPort   uint16 `align:"dport"`
+	sourcePort uint16 `align:"sport"`
+}
```
You can add these to [`datapath/alignchecker/`](https://github.com/cilium/cilium/blob/master/pkg/datapath/alignchecker/alignchecker.go).

qmonnet replied:
Thanks, will do!

pchaigno replied:
While I'm on it: you'll also need to update `bpf/bpf_alignchecker.c`.

pchaigno commented on pkg/datapath/linux/config/config.go:
```diff
@@ -103,6 +104,12 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+
+		if option.Config.EnableIPv4FragmentsTracking {
+			cDefinesMap["IPV4_FRAGMENTS"] = "1"
+			cDefinesMap["IPV4_FRAG_DATAGRAMS_MAP"] = "cilium_ipv4_frag_datagrams"
```
If you export `fragmap.mapName` you could then use it here.

qmonnet replied:
Oh, right, thanks for pointing this out.

pchaigno reviewed (changes requested):
A couple comments on the map creation, on the Go side.

joestringer commented on Documentation/architecture.rst:
```diff
@@ -255,6 +255,7 @@ Load Balancer            node             64k             Max 64k cumulative bac
 Policy                   endpoint         16k             Max 16k allowed identity + port + protocol pairs for specific endpoint
 Proxy Map                node             512k            Max 512k concurrent redirected TCP connections to proxy
 Tunnel                   node             64k             Max 32k nodes (IPv4+IPv6) or 64k nodes (IPv4 or IPv6) across all clusters
+IPv4 Fragmentation       node             8k              Max 8k fragmented datagrams
```
I think that this would be more explicit?

```suggestion
IPv4 Fragmentation       node             8k              Max 8k fragmented datagrams in flight on node
```

qmonnet replied:
Sure, will update :)

joestringer commented on bpf/lib/ipv4.h:
```diff
@@ -62,4 +85,66 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+static __always_inline bool ipv4_is_not_first_fragment(const struct iphdr *ip4)
+{
+	/* Ignore "More fragments" bit to catch all fragments but the first */
+	return ip4->frag_off & bpf_htons(0x9FFF);
```
nit:
```suggestion
	return ip4->frag_off & bpf_htons(0x1FFF);
```

I'm not sure that we should assume any specific meaning from the highest order (reserved) bit here.

qmonnet replied:
This is to remain consistent with [`ipv4_is_fragment()`](https://github.com/cilium/cilium/blob/3713c0076f409fbb98cbb11664706734e4bd953f/bpf/lib/ipv4.h#L75) (“_The evil bit must be set to 0_”),which is called before `ipv4_is_not_first_fragment()`.

I can keep the assumption for both methods, or remove the existing one from `ipv4_is_fragment()` and adapt in my code if you prefer.

joestringer replied:
Ah, it's an interesting case. What does "must be zero" mean in our context of writing packet-forwarding code?

For `ipv4_is_fragment()`, this leads to the drop-packet case. For this codepath, it's leading to the case where we... look up the IP/frag tuple and optionally pass the packet depending on whether there's a match on the tuple and the corresponding port matches policy?

I don't really know the answer. Maybe this is just yak-shaving, in which case maybe we can just disregard my comment.

qmonnet replied:
> Ah, it's an interesting case. What does "must be zero" mean in our context of writing packet-forwarding code?

I understand it as “we expect it to be zero as per the RFC”, if not zero discard the packet.

> For `ipv4_is_fragment()`, this leads to the drop-packet case. For this codepath, it's leading to the case where we... look up the IP/frag tuple and optionally pass the packet depending on whether there's a match on the tuple and the corresponding port matches policy?

Ah, right, I get it. But then with IP fragment support `ipv4_is_fragment()` will no longer be on this drop case (it is called prior to `ipv4_is_not_first_fragment()`).

> I don't really know the answer. Maybe this is just yak-shaving, in which case maybe we can just disregard my comment.

No you're probably right. I think we should take the check on the reserved bit out of `ipv4_is_fragment()` and remove it or move it to its own function (means one more `if`, one more check though :/).

joestringer replied:
Removing it from `ipv4_is_fragment()` seems like the easiest approach here.

qmonnet replied:
Now done in preparatory commit.

joestringer commented on bpf/lib/lb.h:
```diff
@@ -152,13 +153,29 @@ static __always_inline int lb4_select_slave(__u16 count)
 	switch (nexthdr) {
 	case IPPROTO_TCP:
 	case IPPROTO_UDP:
+#if defined LB_4 && defined IPV4_FRAGMENTS && defined HAVE_LRU_MAP_TYPE
```
Is this supposed to be like below?

```suggestion
#if defined LB_L4 && defined IPV4_FRAGMENTS && defined HAVE_LRU_MAP_TYPE
```

qmonnet replied:
:+1:
“Oops”

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -1042,7 +1042,11 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 
 	l4_off = l3_off + ipv4_hdrlen(ip4);
 	csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
+#if !(defined IPV4_FRAGMENTS && defined HAVE_LRU_MAP_TYPE)
```
nit: Throughout the datapath code, there is this double-check for `IPV4_FRAGMENTS` and `HAVE_LRU_MAP_TYPE`, but I think we can skip the second check and only define `IPV4_FRAGMENTS` from userspace if the LRU type is supported?

qmonnet replied:
Yes, I should probably do that. I didn't know how `IPV4_FRAGMENTS` would be passed from user space when I wrote the BPF part.

qmonnet replied:
Updated in new commit, but this required me to add a bpftool probe in pkg/datapath/linux/config/config.go.

joestringer replied:
That should be fine, the probe should be cached after the first time.

joestringer commented on pkg/maps/fragmap/fragmap.go:
```diff
@@ -0,0 +1,109 @@
+	return err
+}
+
+// +k8s:deepcopy-gen=true
```
huh, I always thought that this statement is supposed to appear above a structure. /cc @aanm how does this actually work?

joestringer replied:
(overall question, why here at the end of the file and is there a specific purpose?)

qmonnet replied:
Seems to be a mistake, I think I moved some struct upwards in the file and forgot to move that line (and added it again to the struct). I'll clean it up, thanks for spotting that!

joestringer reviewed (changes requested):
Nice work, looks clean!

I missed the spot where the L4 port actually makes its way into `tuple` for use in policy, was it just my 
oversight or is it not there?

It looks like this is only applying the fragment logic onto the endpoint egress program at the moment (`handle_ipv4_from_lxc`), but I think that the more interesting point will be on ingress towards an endpoint (ie `ipv4_policy()`).

I look forward to seeing what you come up with for the testing of these cases.

For debugging, it could potentially be useful to have a CLI tool to introspect the state of the new table. This could follow as a separate PR though if you like. In a perfect world, we would just use new bpftool + BTF, but I guess that's not guaranteed to be available in the kernels we support today.

The main concern I have here is testing, in particular that currently the CI only tests either an old kernel 4.9 that does not support LRU (and hence would disable this code), or a brand new kernel which has a much smarter verifier. I'm concerned specifically on the first couple of kernels that add LRU support whether the added complexity could trip up the verifier when you try to load this code.

Below is a scattering of minor feedback nits below that I noticed while reviewing.

tgraf commented on bpf/lib/ipv4.h:
```diff
@@ -48,4 +71,66 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+
+	tmp = map_lookup_elem(&IPV4_FRAG_DATAGRAMS_MAP, &frag_id);
+	if (!tmp)
+		return -1;
```
Shouldn't this be a particular `DROP_` reason?

qmonnet replied:
Yes most likely. I'll change it.

tgraf commented on bpf/lib/conntrack.h:
```diff
@@ -468,6 +469,26 @@ static __always_inline void ipv4_ct_tuple_reverse(struct ipv4_ct_tuple *tuple)
+
+	if (unlikely(ipv4_is_fragment(ip4)))
+		return ipv4_handle_fragment(ctx, ip4, off,
+					    (struct ipv4_frag_l4ports *)&tuple->dport);
```
Does this depend on the LB case inserting into the frag map?

qmonnet replied:
It should pick up the fragment registered into the map by the LB (at least, that's the intention). Or register the fragment here if LB hasn't done it (if LB was not on the packet path?).
Does it answer the question?

tgraf commented on pkg/option/config.go:
```diff
@@ -1984,6 +1992,7 @@ func (c *DaemonConfig) Populate() {
 	c.CTMapEntriesTimeoutSVCAny = viper.GetDuration(CTMapEntriesTimeoutSVCAnyName)
 	c.CTMapEntriesTimeoutSYN = viper.GetDuration(CTMapEntriesTimeoutSYNName)
 	c.CTMapEntriesTimeoutFIN = viper.GetDuration(CTMapEntriesTimeoutFINName)
+	c.EnableIPv4FragmentsTracking = viper.GetBool(EnableIPv4FragmentsTrackingName)
```
This still also needs an entry in `daemon/daemon_main.go

qmonnet committed:
bpf: add support for IPv4 fragmentation to CT/policy lookup

Similarly to L4 port extraction for the IPv4 load balancer, add
fragments support for policy checks.

Because the tuple used for L4 (or L3+L4) policy checks is performed as
part of the conntrack lookup, call for L4 ports extraction in
"ct_lookup4()". There are some implications:

- Fragments are not supported for policy checks if "CONNTRACK" is not
  defined.

- Fragment processing is currently performed each time "ct_lookup4()" is
  called for a TCP/UDP packet. This is the case for "ipv4_policy()",
  which we primarily target, but also for "handle_ipv4_from_lxc()",
  "(rev_)nodeport_lb4()", NAT functions, or "lb4_local()". Not sure if
  this is desirable.

If fragments should not be processed on some paths we could deactivate
that with a bool value passed down to "ct_lookup4()". Alternatively, we
can also take port extraction out of there and move it to e.g.
"ipv4_policy()", but we would need a flag to prevent "ct_lookup4()" from
overwriting the collected values. Or we could refactor L4 ports
extraction out of "ct_lookup4()", but that may require more work to make
sure that we keep consistency on all paths.

The need to check whether the fragment is the first for that datagram
also implies that we need to revalidate header data to access the IP
header for each call to "ct_lookup4()". This could maybe be optimised by
looking at the "ip4->id" field before pointers are invalidated, and
passing it down to the function.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
pkg/maps/fragmap: Create IPv4 fragments tracking map in cilium-agent

Create the map used to retrieve the L4 ports asssociated to the datagram
an IPv4 fragment belongs to. It is created by the cilium-agent.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test/K8sServices: test IPv4 datagram fragments support

Add a test to check that fragmented datagrams are correctly routed
through the load balancer and the conntrack lookup.

The test consists in sending fragmented datagrams with netcat from one
"client" pod to a "server" endpoint. To avoid changing the policy in use
for K8sServices, we reuse the existing UDP echo server deployed on the
DS test nodes (although we do not send valid TFTP commands). Reception
of the datagrams is validated by searching the conntrack table for the
correct number of packets for the relevant flow. Using ClusterIP or
NodePort, a number of IP addresses and ports are used to target the
endpoint and validate that the datagram is received in various cases.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: support IPv4 fragments for L4 load balancer key extraction

Add support for IPv4 fragments for L4 load balancer.

Define and use a LRU table for storing information about datagram
fragments. Add the relevant processing functions to ipv4.h. When a
TCP/UDP IPv4 fragment is found:

- If this is the first (logical) fragment of the datagram, add an entry
  to that map using (dst_ip, src_ip, proto, datagram_id) as a key and
  (dst_port, src_port) as value.

- If this is another fragment in the datagram, attempt to retrieve L4
  ports from that map.

Fragment processing is added to "extract_l4_port()", where L4 ports
extraction is performed for load balancing, to avoid adding a new
switch/case on ip4->protocol elsewhere. However, it is worth noting that
"extract_l4_port()", called by "lb(4|6)_extract_key()", is also part of
the "nodeport_lb(4|6)()" path, and I am unsure whether this is
desirable.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: add -DIPV4_FRAGMENTS to BPF Makefile

Add the options for IPv4 frament tracking support to the Makefile, so
that BPF programs are built (and loaded, depending on the case) with the
added complexity caused by fragment tracking support.

Do not add it to MAX_LB_OPTIONS or MAX_LXC_OPTIONS, as the tests
involving those are currently run by CI on Linux 4.9 only (with no LRU
map support - see #10798).

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
pkg/maps/fragmap: add an option to disable IPv4 fragments tracking

Add a configuration option to cilium-agent to offer the possibility to
deactivate support for IPv4 fragments.

The option, "enable-ipv4-fragments-tracking", defaults to true, meaning
the support is on by default. To avoid creating a useless map, it is
disabled (silently) in case IPv4 is not enabled, or (with info-level
message) if LRU maps are not supported.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
bpf: remove check on reserved bit in IPv4 header

In function "ipv4_is_fragment()", we return true if at least one of the
following is verified:

- Flow identifier is non-zero, or
- "More fragments" flag is set, or
- Reserved bit is set.

The reserved bit is not related to fragments. The check is performed
here because if the bit is set, then the packet is recognised as a
fragment, thus leading to the BPF program discarding it.

However, future changes may add support for IPv4 datagram fragments. In
such case, it is undesirable to assimilate the packet to a fragment if
the reserved bit only is set.

The check on the reserved bit should ideally be moved out of the
"ipv4_is_fragment()". But to avoid adding additional checks elsewhere in
the datapaths, this commit simply removes the check. Packets are now
recognised as fragments if at least one of the first two conditions is
met.

Suggested-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno commented on pkg/maps/fragmap/fragmap.go:
```diff
@@ -0,0 +1,92 @@
+		0,
+		bpf.ConvertKeyValue,
+	)
+	_, err := fragMap.OpenOrCreate()
```
I think this should be:
```suggestion
	err := fragmap.Create()
```
We're not using this map in the daemon so we can close the fd.

qmonnet replied:
Ok. I kept it open mostly to be able to see it with `cilium map list`, for debug. But I guess this is a weak argument to keep it open, I can close it.

pchaigno replied:
I don't have a strong opinion either way, but if we do that, we need to do it for all other similar maps I guess, for consistency.

qmonnet replied:
Sure, I'll go for consistency with this PR, and close the fd. We can discuss opening it for the maps in general during the next datapath call, to see if there's an interest in doing a follow-up on this.

pchaigno commented on daemon/cmd/datapath.go:
```diff
@@ -402,6 +403,10 @@ func (d *Daemon) initMaps() error {
 		}
 	}
 
+	if err := fragmap.InitMap(); err != nil {
```
Shouldn't this map only be created if LRU maps are supported? You check this condition for the BPF program generation, but I haven't seen such a check on the Go side.

qmonnet replied:
The check is in the `fragmap` package:
``` go
func InitMap() error {
	if !option.Config.EnableIPv4 || !option.Config.EnableIPv4FragmentsTracking {
		return nil
	}
```

`option.Config.EnableIPv4FragmentsTracking` is not enabled if LRU are not supported.

pchaigno replied:
> `option.Config.EnableIPv4FragmentsTracking` is not enabled if LRU are not supported.

It's not? Where does that happen? I've seen that you print a warning if LRU maps are not supported; maybe you forgot to disable `EnableIPv4FragmentsTracking` at the same time?

qmonnet replied:
Ah, right, maybe I did forget :open_mouth:. Looks like I disabled the fragment-related bits for BPF programs only :(. Will fix, thanks!

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -159,13 +160,30 @@ static __always_inline int lb4_select_slave(__u16 count)
 	case IPPROTO_TCP:
 	case IPPROTO_UDP:
+#ifdef IPV4_FRAGMENTS
+		if (is_ipv4) {
```
Wouldn't checking `if (ip4)` be enough here?

qmonnet replied:
But then we would run it even if we do not have LRU support?

pchaigno replied:
I mean using `if (ip4)` instead of `if (is_ipv4)`. I'm not suggesting we remove `#ifdef IPV4_FRAGMENTS` ;-)

qmonnet replied:
Ah, sorry. To remove the `is_ipv4` argument then? Right, I'll do that, thanks.

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -466,6 +467,26 @@ static __always_inline void ipv4_ct_tuple_reverse(struct ipv4_ct_tuple *tuple)
+
+	// TODO: Can we optimise this out?
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
```
Why is it needed? Wasn't the IPv4 header already validated to be able to load the sport+dport? And `off` already points to the IPv4 header, no?

qmonnet replied:
This is called from `ct_lookup4()`, and bpf_lxc.c has the following warning just before calling that function from `handle_ipv4_from_lxc()`: 
``` c
	/* WARNING: ip4 offset check invalidated, revalidate before use */
```

pchaigno replied:
Ah, I see. This check wasn't needed before because `ctx_load_bytes` has it's own length check (even in XDP case).

I think the TODO can be removed (unless you had an idea in mind to optimize it out?).

qmonnet replied:
Not really, not without taking L4 ports extraction out of `ct_lookup4()`, but that would probably require quite some amount of refactoring. I'll drop the TODO, maybe replace by a comment to tell why revalidation is required.

pchaigno reviewed (changes requested):
A few comments below. I haven't checked the test yet. Also not sure if we need any new metrics associated with IP fragments.

aanm commented on pkg/datapath/linux/config/config.go:
```diff
@@ -105,6 +107,17 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+			cDefinesMap["CILIUM_IPV4_FRAG_MAP_MAX_ENTRIES"] = fmt.Sprintf("%d", fragmap.MaxEntries)
+			// } else {
+			// TODO: Log something? But "log" not imported in this file yet
+			// log.Warningf("No support for IPv4 fragments (LRU maps not supported by kernel)")
```
This logging can be performed when we are initializing the daemon, around the same place where we perform the bpf probing. For example here: https://github.com/cilium/cilium/blob/e1c05e24814c5aff774305836b2c2e825793b272/pkg/datapath/linux/requirements.go#L175-L184

aanm commented on pkg/datapath/linux/config/config.go:
```diff
@@ -31,6 +31,8 @@ import (
 	"github.com/cilium/cilium/pkg/defaults"
 	"github.com/cilium/cilium/pkg/identity"
```
If we opt by me comment left in the previous commit this commit can be dropped.

qmonnet replied:
That's a good idea, I will do as you suggest and drop this commit.

aanm reviewed (commented):
Changes look good to me for the golang bits. Great work @qmonnet

pchaigno commented on bpf/lib/conntrack.h:
```diff
@@ -466,6 +467,26 @@ static __always_inline void ipv4_ct_tuple_reverse(struct ipv4_ct_tuple *tuple)
+					    (struct ipv4_frag_l4ports *)&tuple->dport);
+#endif
+	/* load sport + dport into tuple */
+	return ctx_load_bytes(ctx, off, &tuple->dport, 4);
```
Couldn't this use direct packet copy?

qmonnet replied:
For this one though, that would require doing the `revalidate_data()` even if `IPV4_FRAGMENTS` is not set?

pchaigno replied:
Ah, right. Probably best to leave it that way until we rework all the direct packet loads then.

pchaigno commented on bpf/lib/ipv4.h:
```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+{
+	int ret;
+
+	ret = ctx_load_bytes(ctx, l4_off, ports, 4);
```
Same here for direct packet copy. Plus, it looks like it would remove the need to pass `l4_off` as an argument to a few functions.

qmonnet replied:
Right, makes sense. I'll switch to direct access.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -105,6 +111,18 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
+				cDefinesMap["IPV4_FRAG_DATAGRAMS_MAP"] = fragmap.MapName
+				cDefinesMap["CILIUM_IPV4_FRAG_MAP_MAX_ENTRIES"] = fmt.Sprintf("%d", fragmap.MaxEntries)
+			} else {
+				log.Warningf("No support for IPv4 fragments (LRU maps not supported by kernel)")
```
Do we need this to be `Warning` or should it just be `Info`? People follow metrics / set alerts on the number of warnings we report from the daemon. Also, the `XXXf` form is not necessary here as there are no additional arguments. We typically try to avoid this form in favour of `log.WithFields(...).Info()`.

```suggestion
				log.Info("No support for IPv4 fragments (LRU maps not supported by kernel)")
```

qmonnet replied:
I already noticed and replaced the `XXXf` form with `Warning()`. But I don't think we _need_ a warning here, so I'll just go with `Info()`.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -50,6 +54,8 @@ import (
 	"github.com/vishvananda/netlink"
 )
 
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "datapath-config")
```
Is it possible to move the logging part of this into options validation in the daemon?

Inviting the logging into the config package here increases the likelihood that we end up re-logging the same message again and again any time we make datapath changes. In the node case like this, it'll only be if someone reconfigures the daemon I think; but if we end up adding log messages to the endpointconfig / netdev config then it will print every time someone starts an endpoint, and potentially multiple times as regenerations are requested.

qmonnet replied:
> Is it possible to move the logging part of this into options validation in the daemon?

Moved to pkg/datapath/linux/requirements.go, as [suggested by André](https://github.com/cilium/cilium/pull/10264#discussion_r401425558).

> Inviting the logging into the config package here increases the likelihood that we end up re-logging the same message again and again any time we make datapath changes. In the node case like this, it'll only be if someone reconfigures the daemon I think; but if we end up adding log messages to the endpointconfig / netdev config then it will print every time someone starts an endpoint, and potentially multiple times as regenerations are requested.

I understand, I thought adding the log package was maybe not desirable (hence the separate commit).

qmonnet replied:
> Moved to pkg/datapath/linux/requirements.go

On second thought, moving to `initEnv()`.

joestringer commented on pkg/maps/fragmap/fragmap.go:
```diff
@@ -0,0 +1,92 @@
+		0,
+		bpf.ConvertKeyValue,
+	)
+	_, err := fragMap.OpenOrCreate()
```
From `OpenOrCreate` comment:

https://github.com/cilium/cilium/blob/26dec4c4f4311df2b1a6c909b27ff7fe6e46929f/pkg/bpf/map_linux.go#L428-L432

Did I miss it or will this always create the map as type hashmap, even if we end up disabling this functionality because we run on an older kernel?

joestringer replied:
I think if we parse the option in the daemon package and disable the option globally when LRU is not supported, this problem will resolve itself because of the check earlier in this function.

qmonnet replied:
The idea is that the `InitMap()` function should not be called if LRU maps are not supported. Paul highlighted that I forgot the relevant check, so the fix is coming.

joestringer commented on test/k8sT/Services.go:
```diff
@@ -846,6 +986,13 @@ var _ = Describe("K8sServicesTest", func() {
 			})
+
+		// Net-next and not old versions, because of LRU requirement.
+		// TODO: Skip if no IPv4 or if fragment support is disabled
```
These services tests I think currently always enable IPv4, and disabling fragments support is not currently configured in CI. Are there any actionable items on this TODO?

qmonnet replied:
I wasn't sure, hence the TODO. But if we're good with the code and CI doesn't have a chance to run without IPv4/fragment option, I'll remove the comment.

joestringer commented on bpf/lib/ipv4.h:
```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+
+	tmp = map_lookup_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id);
+	if (!tmp)
+		return DROP_FRAG_NOSUPPORT;
```
Do you think we should use a different error code, eg `DROP_FRAG_NOT_FOUND` to differentiate this error vs. lack of support?

I can't decide whether this is a good idea or not. Technically we can determine whether this error code means that we don't support fragments or if it means that the fragment info was not found (eg out-of-order) because we will only log the message for this one either when fragment support is disabled for any fragment, or when fragment support is enabled with out-of-order / spurious random fragments not seen before. Would be cognitively easier to understand at a glance if at least the `cilium monitor` message made it clear which of these two cases are occurring though. Less digging through current state to understand.

qmonnet replied:
Makes sense, I thought of adding a new code but wasn't sure. As you say, it will probably make things more explicit, I'll add one.

joestringer commented on test/k8sT/Services.go:
```diff
@@ -688,6 +757,77 @@ var _ = Describe("K8sServicesTest", func() {
+			// TODO: We want TargetPort but it is of type
+			// intstr.IntOrString, so the following does not work.
+			// Not sure how to unwrap it.
+			//serverPort := data.Spec.Ports[1].TargetPort
```
Seems like this should just require `IntValue()`?

https://godoc.org/k8s.io/apimachinery/pkg/util/intstr#IntOrString.IntValue

qmonnet replied:
Oh, looks like the function I was looking for. Thank you for that, and for the review!

joestringer reviewed (changes requested):
This looks awesome, super close.

A few minor comments below.

qmonnet committed:
bpf: add new DROP_FRAG_NOT_FOUND error code

The new DROP_FRAG_NOT_FOUND error code is returned when IPv4 fragment
tracking is enabled, but we were not able to find an entry for the
datagram to which the current packet belongs in the relevant map. This
would typically occur when datagram fragments do not arrive in the
correct order, and we have not processed the first logical fragment
(with L4 header) to create the relevant entry and register L4 ports for
that datagram.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

pchaigno commented on bpf/Makefile:
```diff
@@ -58,7 +59,7 @@ LB_OPTIONS = \
 MAX_LB_OPTIONS = -DENABLE_IPV4 -DENABLE_IPV6 -DLB_L3 -DLB_L4 \
 	-DENABLE_HOST_SERVICES_TCP -DENABLE_HOST_SERVICES_UDP \
-	-DENABLE_IPSEC
+	-DENABLE_IPSEC -DIPV4_FRAGMENTS
```
You can't enable IPv4 fragments here because this set of options is tested on 4.9. This is likely the reason why the `RuntimeVerifier` test currently fails.

Related: #10798.

qmonnet replied:
Oh I missed that. Ok, will remove until that issue is fixed. Do you know if leaving the directive in `LB_OPTIONS` and `LXC_OPTIONS` is safe?

pchaigno replied:
Yes, it should be safe. Only the `MAX_*_OPTIONS` are used for the `RuntimeVerifier` test.

pchaigno commented on bpf/Makefile:
```diff
@@ -151,7 +153,7 @@ LXC_OPTIONS = \
 # tested as well.
 MAX_LXC_OPTIONS = -DENABLE_IPV4 -DENABLE_IPV6 -DENABLE_HOST_REDIRECT -DENABLE_NAT46 \
-	-DENABLE_ROUTING -DENABLE_IPSEC
+	-DENABLE_ROUTING -DENABLE_IPSEC -DIPV4_FRAGMENTS
```
Ditto.

tklauser commented on pkg/datapath/maps/map.go:
```diff
@@ -163,6 +163,11 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
 
+	if !option.Config.EnableIPv4FragmentsTracking {
+		maps = append(maps, []string{
+			"cilium_ipv4_frag_datagrams"}...)
```
Nit: this can be simplified to:

```suggestion
		maps = append(maps, "cilium_ipv4_frag_datagrams")
```

qmonnet replied:
Thanks, I'm pushing the fix

joestringer commented on bpf/lib/common.h:
```diff
@@ -332,6 +332,7 @@ enum {
 #define DROP_UNKNOWN_SENDER	-172
 #define DROP_NAT_NOT_NEEDED	-173 /* Mapped as drop code, though drop not necessary. */
 #define DROP_IS_CLUSTER_IP	-174
+#define DROP_FRAG_NOT_FOUND	-175
```
I think this needs the corresponding formatting side in `pkg/monitor/datapath_drop.go`.

qmonnet replied:
I thought more work was needed to add new metrics and that I would do a follow-up PR, but adding return code and reason are apparently enough, so sure I'll add it!

joestringer commented on daemon/cmd/datapath.go:
```diff
@@ -403,6 +404,12 @@ func (d *Daemon) initMaps() error {
 	}
 
+	if !option.Config.EnableIPv4FragmentsTracking {
+		if err := fragmap.InitMap(); err != nil {
```
Is this meant to be reversed?

```suggestion
	if option.Config.EnableIPv4FragmentsTracking {
		if err := fragmap.InitMap(); err != nil {
```

joestringer replied:
In practice this map will be created during the datapath load anyway so it's not critical which probably explains why the tests can pass, but good to fix up.

qmonnet replied:
Ouch, nice catch! Thanks.

joestringer commented on test/k8sT/Services.go:
```diff
@@ -439,6 +439,127 @@ var _ = Describe("K8sServicesTest", func() {
+			// packets.
+
+			// Field #7 is "RxPackets=<n>"
+			cmdIn := "cilium bpf ct list global | awk '/%s/ { sub(\".*=\",\"\", $7); print $7 }'"
```
We have JSON encoding to make this kind of call more resilient.

The CT list can use:
```
cilium bpf ct list global -o json
```

I believe that if this is just filtering the RxPackets field and *only* printing that, the equivalent jq filter is:

```
jq '.MapValue.RxPackets'
```

This is not blocking this PR but it would be nice to follow up after this PR to fix this kind of case up just in case we change the standard CLI format in future. Same for the other calls in this test.

qmonnet replied:
Right, I didn't realise we had JSON here.

... On the other hand, JSON output for ct map is currently broken (#2976). Oh well. I'll fix in a separate PR, and update here as a follow-up. I saw cilium also has a `-o jsonpath` option, but it seems less flexible than `jq`, I'll see what works best when doing the follow-up.

joestringer replied:
Thanks for digging into this, I didn't realise it was still spitting out the wrong output.

re: jsonpath vs. jq, I didn't think `-o jsonpath` was supported for `cilium bpf ct list`. We use jsonpath a lot in the tests and unfortunately the syntax isn't interchangeable but *shrug* whatever works for you IMO.

joestringer reviewed (changes requested):
Super close. Please address the first two comments below in this PR; the tests fixup I don't mind deferring to a subsequent follow-up PR.

borkmann reviewed (approved):
Just small nits I can comment later on, great work!

borkmann commented on Documentation/architecture.rst:
```diff
@@ -255,6 +255,7 @@ Load Balancer            node             64k             Max 64k cumulative bac
 Policy                   endpoint         16k             Max 16k allowed identity + port + protocol pairs for specific endpoint
 Proxy Map                node             512k            Max 512k concurrent redirected TCP connections to proxy
 Tunnel                   node             64k             Max 32k nodes (IPv4+IPv6) or 64k nodes (IPv4 or IPv6) across all clusters
+IPv4 Fragmentation       node             8k              Max 8k fragmented datagrams in flight on node
```
nit: `Max 8k fragmented datagrams in flight simultaneously on the node`?

borkmann replied:
Should we make this size configurable? So a user can configure a table of more than 8k?

qmonnet replied:
Will address the nit. For the configurable size, I don't know. We don't expect many IPv4 fragments in the first place, so the reasoning was that 8k should be probably be enough. Should we wait and see if there is demand for more entries, or would you prefer me to add an option now?

borkmann replied:
I presume in normal case it might be only DNS related or not used at all, unless a user misconfigures their network. Maybe in that sense 1k would even be enough as a default as we prealloc the whole map? Given we have tunables for `bpf-policy-map-max` for example, it would probably make sense to add one here especially if we have it on a lower size by default.

qmonnet replied:
Ok, I labelled the follow-up PR as WIP, will add a commit to introduce a tunable here as well.

borkmann commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -61,6 +61,7 @@ cilium-agent [flags]
       --enable-hubble                                 Enable hubble server
       --enable-ipsec                                  Enable IPSec support
       --enable-ipv4                                   Enable IPv4 support (default true)
+      --enable-ipv4-fragments-tracking                Enable IPv4 fragments tracking for L4-based lookups (default true)
```
nit: s/fragments-tracking/fragment-tracking/ ?

borkmann commented on bpf/bpf_lxc.c:
```diff
@@ -1032,7 +1032,12 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, __u8 *reason,
 	l4_off = l3_off + ipv4_hdrlen(ip4);
 	csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
-	is_fragment = ipv4_is_fragment(ip4);
+#if !defined IPV4_FRAGMENTS
```
nit: `#ifndef ENABLE_IPV4_FRAGMENTS`; we usually have it as `ENABLE_` prefix for such configs

Throughout the code we also use `defined(...)` and not `defined ...`.

qmonnet replied:
There seems to be many `defined ...` too under the bpf/ directory. But sure, no problem! Will change to `#ifdef`/`#ifndef`. And I'll add the `ENABLE_` prefix.

borkmann replied:
Right, the `defined ...` one is probably worth a bpf/ wide cleanup to make it consistent.

qmonnet replied:
Will add to WIP follow-up PR.

borkmann commented on bpf/lib/ipv4.h:
```diff
@@ -8,6 +8,30 @@
+	__be16	id;		/* L4 datagram identifier */
+	__u8	proto;
+	__u8	pad;
+} __attribute__((packed));
```
nit: we should have __packed available as macro.

qmonnet replied:
We do, although it is not used in the code at the moment. And it didn't exist when I pushed this code to the PR :astonished:. Sure, will change!

borkmann replied:
Oh, I wasn't aware, no worries then. :-) Probably another one for bpf/ wide cleanup to convert all `__attribute__((packed))` into `__packed`. If you prefer I could also do the packed and define one later for all bpf/, up to you.

qmonnet replied:
I'll do the cleanup.

borkmann commented on bpf/lib/ipv4.h:
```diff
@@ -8,6 +8,30 @@
+	__be16	dport;
+} __attribute__((packed));
+
+#if defined IPV4_FRAGMENTS
```
nit: `#ifdef ...`

borkmann commented on bpf/lib/ipv4.h:
```diff
@@ -8,6 +8,30 @@
+	.size_value	= sizeof(struct ipv4_frag_l4ports),
+	.pinning	= PIN_GLOBAL_NS,
+	.max_elem	= CILIUM_IPV4_FRAG_MAP_MAX_ENTRIES,
+	.flags		= CONDITIONAL_PREALLOC,
```
Does LRU hash even work with no prealloc?

qmonnet replied:
kernel/bpf/hashtab.c says:
``` c
	if (lru && !prealloc)
		return -ENOTSUPP;
```
So, no. Uh, I probably copied this flag and forgot to check its meaning :fearful:. Let's remove it quick \<insert parrot>!

borkmann commented on bpf/lib/ipv4.h:
```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+		return DROP_FRAG_NOT_FOUND;
+
+	/* Do not make ports a pointer to map data, copy from map */
+	__builtin_memcpy(ports, tmp, sizeof(*ports));
```
We can use just `memcpy()`. This will use the __builtin_memcpy() internally.

borkmann commented on bpf/lib/ipv4.h:
```diff
@@ -44,8 +68,71 @@ static __always_inline bool ipv4_is_fragment(struct iphdr *ip4)
+	if (ret < 0)
+		return ret;
+
+	map_update_elem(&IPV4_FRAG_DATAGRAMS_MAP, frag_id, ports, BPF_ANY);
```
Can we add a metric to the Cilium metrics map to have some introspection that fragments were seen? Would be super useful to have this visible in Prometheus as a graph later to see when/if potential configuration issues appeared.

borkmann replied:
Maybe also a metric if the map update failed to be able to debug that case when it could not be handled.
